File: q100-q199/q110-109/q111/q111.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
import queue
class Solution:
    def minDepth(self, root: TreeNode) -> int:
        if root == None:
            return 0
        level_que = queue.SimpleQueue()
        level_que.put(root)
        min_depth = 0
        while not level_que.empty():
            level_size = level_que.qsize()
            min_depth += 1
            for i in range(0, level_size):
                node = level_que.get()
                if node.left == None and node.right == None:
                    return min_depth
                if node.left != None:
                    level_que.put(node.left)
                if node.right != None:
                    level_que.put(node.right)
        return min_depth

File: q100-q199/q110-109/q111/q111.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int minDepth(TreeNode* root) {
        if (root == NULL) {
            return 0;
        }
        queue<TreeNode*> level_que;
        int min_depth = 0;
        level_que.push(root);
        while (!level_que.empty()) {
            int level_size = level_que.size();
            ++min_depth;
            for (int i = 0; i < level_size; ++i) {
                TreeNode* node = level_que.front();
                level_que.pop();
                if (node->left == NULL && node->right == NULL) {
                    return min_depth;
                }
                if (node->left != NULL) {
                    level_que.push(node->left);
                }
                if (node->right != NULL) {
                    level_que.push(node->right);
                }
            }
        }
        return min_depth;
    }
};
